File: physical-ai-textbook/docs/chapter-2/assets/code-examples/talker.py
#!/usr/bin/env python3
"""
ROS 2 Talker Node (Publisher Example)

This node publishes String messages to the '/chatter' topic at 10Hz.

Implementation: T025 (Phase 3: User Story 1)

Usage:
    ros2 run <package_name> talker
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class TalkerNode(Node):
    def __init__(self):
        super().__init__('talker')

        # Create publisher: topic='/chatter', message type=String, queue size=10
        self.publisher = self.create_publisher(String, '/chatter', 10)

        # Create timer: 10Hz (0.1 seconds)
        self.timer = self.create_timer(0.1, self.timer_callback)

        # Message counter
        self.counter = 0

        self.get_logger().info('Talker node started')

    def timer_callback(self):
        """Publish a message every timer tick"""
        msg = String()
        msg.data = f'Hello World: {self.counter}'

        self.publisher.publish(msg)
        self.get_logger().info(f'Publishing: "{msg.data}"')

        self.counter += 1


def main(args=None):
    # Initialize ROS 2 Python client library
    rclpy.init(args=args)

    # Create node instance
    node = TalkerNode()

    try:
        # Spin node (process callbacks)
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        # Clean shutdown
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: physical-ai-textbook/docs/chapter-2/assets/code-examples/talker.cpp
/**
 * ROS 2 Talker Node (Publisher Example) - C++ Version
 *
 * This node publishes String messages to the '/chatter' topic at 10Hz.
 *
 * Implementation: T027 (Phase 3: User Story 1)
 *
 * Usage:
 *     ros2 run <package_name> talker_cpp
 *
 * Compilation (CMakeLists.txt):
 *     find_package(rclcpp REQUIRED)
 *     find_package(std_msgs REQUIRED)
 *     add_executable(talker_cpp src/talker.cpp)
 *     ament_target_dependencies(talker_cpp rclcpp std_msgs)
 */

#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class TalkerNode : public rclcpp::Node
{
public:
    TalkerNode() : Node("talker"), counter_(0)
    {
        // Create publisher: topic='/chatter', queue size=10
        publisher_ = this->create_publisher<std_msgs::msg::String>("/chatter", 10);

        // Create timer: 10Hz (100ms)
        timer_ = this->create_wall_timer(
            100ms,
            std::bind(&TalkerNode::timer_callback, this)
        );

        RCLCPP_INFO(this->get_logger(), "Talker node started");
    }

private:
    void timer_callback()
    {
        auto message = std_msgs::msg::String();
        message.data = "Hello World: " + std::to_string(counter_);

        publisher_->publish(message);
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());

        counter_++;
    }

    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    size_t counter_;
};

int main(int argc, char * argv[])
{
    // Initialize ROS 2 C++ client library
    rclcpp::init(argc, argv);

    // Create node instance
    auto node = std::make_shared<TalkerNode>();

    // Spin node (process callbacks)
    rclcpp::spin(node);

    // Clean shutdown
    rclcpp::shutdown();
    return 0;
}
